File: sections/74_geometric_sweeps_proximity/problems/d_rectangle_union_perimeter/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;

struct Event {
    int64 x, y1, y2;
    int delta;
};

struct CoverTree {
    int intervals;
    const vector<int64>& coordinate;
    vector<int> cover, components;
    vector<int64> length;
    vector<char> left_covered, right_covered;

    explicit CoverTree(const vector<int64>& ys)
        : intervals((int)ys.size() - 1), coordinate(ys),
          cover(4 * max(1, intervals)), components(cover.size()),
          length(cover.size()), left_covered(cover.size()),
          right_covered(cover.size()) {}

    void pull(int node, int low, int high) {
        if (cover[node] > 0) {
            length[node] = coordinate[high] - coordinate[low];
            components[node] = 1;
            left_covered[node] = right_covered[node] = true;
        } else if (high - low == 1) {
            length[node] = 0;
            components[node] = 0;
            left_covered[node] = right_covered[node] = false;
        } else {
            int left = 2 * node;
            int right = left + 1;
            length[node] = length[left] + length[right];
            components[node] = components[left] + components[right]
                             - (right_covered[left] && left_covered[right]);
            left_covered[node] = left_covered[left];
            right_covered[node] = right_covered[right];
        }
    }

    void update(int from, int to, int delta, int node, int low, int high) {
        if (to <= low || high <= from) return;
        if (from <= low && high <= to) {
            cover[node] += delta;
            pull(node, low, high);
            return;
        }
        int middle = (low + high) / 2;
        update(from, to, delta, 2 * node, low, middle);
        update(from, to, delta, 2 * node + 1, middle, high);
        pull(node, low, high);
    }

    void update(int from, int to, int delta) {
        update(from, to, delta, 1, 0, intervals);
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<Event> events;
    vector<int64> ys;
    for (int i = 0; i < n; ++i) {
        int64 x1, y1, x2, y2;
        cin >> x1 >> y1 >> x2 >> y2;
        events.push_back({x1, y1, y2, +1});
        events.push_back({x2, y1, y2, -1});
        ys.push_back(y1);
        ys.push_back(y2);
    }
    sort(ys.begin(), ys.end());
    ys.erase(unique(ys.begin(), ys.end()), ys.end());
    sort(events.begin(), events.end(), [](const Event& a, const Event& b) {
        if (a.x != b.x) return a.x < b.x;
        return a.delta > b.delta;  // additions before removals
    });

    CoverTree tree(ys);
    int64 area = 0;
    int64 perimeter = 0;
    int64 previous_x = events.front().x;
    int index = 0;
    while (index < (int)events.size()) {
        int64 x = events[index].x;
        int64 width = x - previous_x;
        area += tree.length[1] * width;
        perimeter += int64(2) * tree.components[1] * width;

        while (index < (int)events.size() && events[index].x == x) {
            int from = lower_bound(ys.begin(), ys.end(), events[index].y1)
                     - ys.begin();
            int to = lower_bound(ys.begin(), ys.end(), events[index].y2)
                   - ys.begin();
            int64 before = tree.length[1];
            tree.update(from, to, events[index].delta);
            perimeter += llabs(tree.length[1] - before);
            ++index;
        }
        previous_x = x;
    }

    cout << area << ' ' << perimeter << '\n';
}

File: sections/74_geometric_sweeps_proximity/problems/d_rectangle_union_perimeter/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    events = []
    ys = []
    at = 1
    for _ in range(n):
        x1, y1, x2, y2 = data[at : at + 4]
        at += 4
        events.append((x1, -1, y1, y2))  # additions sort first
        events.append((x2, +1, y1, y2))
        ys.extend((y1, y2))
    ys = sorted(set(ys))
    position = {value: index for index, value in enumerate(ys)}
    events.sort()

    size = 4 * (len(ys) - 1)
    cover = [0] * size
    length = [0] * size
    components = [0] * size
    left_covered = [False] * size
    right_covered = [False] * size

    def pull(node: int, low: int, high: int) -> None:
        if cover[node] > 0:
            length[node] = ys[high] - ys[low]
            components[node] = 1
            left_covered[node] = right_covered[node] = True
        elif high - low == 1:
            length[node] = components[node] = 0
            left_covered[node] = right_covered[node] = False
        else:
            left = 2 * node
            right = left + 1
            length[node] = length[left] + length[right]
            components[node] = (
                components[left] + components[right]
                - (right_covered[left] and left_covered[right])
            )
            left_covered[node] = left_covered[left]
            right_covered[node] = right_covered[right]

    def update(start: int, end: int, delta: int, node=1, low=0, high=None) -> None:
        if high is None:
            high = len(ys) - 1
        if end <= low or high <= start:
            return
        if start <= low and high <= end:
            cover[node] += delta
            pull(node, low, high)
            return
        middle = (low + high) // 2
        update(start, end, delta, 2 * node, low, middle)
        update(start, end, delta, 2 * node + 1, middle, high)
        pull(node, low, high)

    area = perimeter = 0
    previous_x = events[0][0]
    index = 0
    while index < len(events):
        x = events[index][0]
        width = x - previous_x
        area += length[1] * width
        perimeter += 2 * components[1] * width
        while index < len(events) and events[index][0] == x:
            _, encoded_delta, y1, y2 = events[index]
            before = length[1]
            update(position[y1], position[y2], -encoded_delta)
            perimeter += abs(length[1] - before)
            index += 1
        previous_x = x

    print(area, perimeter)


if __name__ == "__main__":
    main()
